### ISSUE TITLE
Smallest possible dependencies for a grpc-java client

### ISSUE THREAD

**Issue by nadundesilva** [grpc/grpc-java#7107]
I am trying to use grpc-java to implement a grpc client. I was able to implement it using the Getting Started guide.

However, in my case, the size of the compiled shaded Jar of my program plays a very crucial role (even reducing by few MBs would be helpful). Therefore I was wondering what the smallest possible dependencies that I could use. I am currently using the following dependencies.
```
implementation 'io.grpc:grpc-api:1.27.2'
implementation 'io.grpc:grpc-core:1.27.2'
implementation 'io.grpc:grpc-netty-shaded:1.27.2'
implementation 'io.grpc:grpc-protobuf:1.27.2'
implementation 'io.grpc:grpc-stub:1.27.2'
```

I noticed how `grpc-okhttp` and `grpc-protobuf-lite` is recommended for Android clients. Can I use them (since they seems to be very low in size) although my client will be running in a non-android environment ? What are the differences between the implementations ? (I found out that protobuf-lite lacks some capabilities, although I am not sure exactly what are lacking)

**Comment by voidzcy** [comment]
Yes, you can definitely use okhttp and protobuf-lite on non-Android environment. okhttp is a light wight transport that is particularly suitable for applications running on resource-limited environments. It is intended to be something simple comparing with Netty, while Netty is intended to be more powerful and performant under high throughput use cases.

protobuf-lite has a much smaller code size compared to normal protobuf. It's a separated design and implementation. You may want to find discussions online for the difference between them. But for gRPC's usage, protobuf-lite has all the capabilities as protobuf.

There isn't much change for the client application to swap to okhttp and protobuf-lite. You can simply follow an Android example. It still works on non-Android environments.

**Comment by nadundesilva** [comment]
@voidzcy Thank you! That solves the question I had.

Just a suggestion. Is there any reason for recommending okhttp for Android ? If not we can maybe rephrase it in the Getting Started Guide.

**Comment by ejona86** [comment]
I would not suggest using protobuf-lite, unless the protobuf descriptors included with full protobuf (used for reflection) produce too much code bloat for you. I'd only recommend using protobuf-lite if you care enough about code bloat that' you'd be using a tool that does dead code elimination (like Proguard). grpc-okhttp is easier to recommend, as it has fewer sharp edges compared to protobuf-lite, but just realize it is optimized for a different environment than you will be using it in (Android and OpenJDK have substantially different performance when it comes to blocking vs nonblocking I/O, allocation cost, compiler optimizations, lifetime of processes, and implicitly networking environments).

Note that Netty doesn't have to be as large as it is. It is the current size to support things like Java 8 in a semi-cross-platform way. If running on Java 8 and using grpc-okhttp, you'll need Conscrypt which is also MB in size. (Although see https://github.com/grpc/grpc-java/issues/6997 where with Java 8u252 we'll eventually be able to avoid this dependency. But note that throughput will be limited to 10% of what you'd see with these binaries, because Java 8 AES GCM crypto performance was _poor_.)

Both netty and okhttp can be made smaller if you know which specific platform you will run on, like x86_64 Linux, as you can include just the one binary necessary for that one platform. Normally we include binaries for Linux, Window, and OS X.

If running on Java 11, you can use just grpc-okhttp or just grpc-netty, without any need for the large binaries. We typically still recommend using the binaries in these environments, because Java security performance and feature uptake is generally too slow, but in Java 11 it is "reasonable."

Most of this is discussed in our [SECURITY.md](https://github.com/grpc/grpc-java/blob/master/SECURITY.md) documentation. For example, the end of the "[TLS with netty-tcnative on BoringSSL](https://github.com/grpc/grpc-java/blob/master/SECURITY.md#tls-with-netty-tcnative-on-boringssl)" section mentions how to depend on the precise platform's security binary.

**Comment by ejona86** [comment]
Seems like this is answered. If not, comment, and it can be reopened.
